wavedump/src/fit-histograms

```diff
@@ -0,0 +1,566 @@
+#!/usr/bin/env python3
+from __future__ import print_function, division
+import h5py
+import numpy as np
+import sys
+import csv
+import os
+import ROOT
+from ROOT import gROOT
+from ROOT import TMath
+
+# ROOT.ROOT.EnableThreadSafety()
+
+# DEFAULT VALUES FOR SPE FIT:
+D_OFFSET = 0
+D_LAMBDA = 1
+D_SPE_CHARGE = 0.8
+D_NOISE_SPREAD = 0.01
+D_SPE_CHARGE_SPREAD = 0
+D_ZERO_PEAK_SPREAD = 0.4
+# Number of peaks we use for SPE fitting
+n = 20
+
+def iqr(h):
+    """
+    Returns the iqr of the histogram `h`
+    """
+    l = None
+    r = None
+    total = h.GetEntries()
+    i = 0
+    n = h.GetBinContent(i)
+    while n / total < 0.25:
+        i += 1
+        n += h.GetBinContent(i)
+    high_prob = n / total
+    n -= h.GetBinContent(i)
+    i -= 1
+    low_prob = n / total
+    l = (((0.25 - low_prob) / (high_prob - low_prob)) * h.GetBinWidth(i)) + h.GetBinLowEdge(i)
+    while n / total < 0.75:
+        i += 1
+        n += h.GetBinContent(i)
+    high_prob = n / total
+    n -= h.GetBinContent(i)
+    i -= 1
+    low_prob = n / total
+    r = (((0.75 - low_prob) / (high_prob - low_prob)) * h.GetBinWidth(i)) + h.GetBinLowEdge(i)
+    return r-l
+
+def average_distance_from(val, h):
+    """
+    Standard deviation calculation, except use `val` instead of the mean.
+    """
+    d2 = 0
+    for i in range(h.GetMaximumBin() + 2):
+        d2 += (val - h.GetXaxis().GetBinCenter(i))**2 * h.GetBinContent(i)
+    return np.sqrt(d2)
+
+def save_as_csv(data_type, value, value_err, channel):
+    """
+    Saves charge data in `temp_data.csv`.
+    `temp_data.csv` is read by `save_spe_data.py`, which
+    organizes the data and saves it to `Fit_Data.csv`.
+    """
+    with open ('temp_data.csv', 'a') as file:
+        headers = [
+            'Data Type',
+            'Channel',
+            'Charge',
+            'Err Abs'
+            ]
+        writer = csv.DictWriter(file, headers)
+        if os.stat('temp_data.csv').st_size == 0:
+            writer.writeheader()
+        writer.writerow({
+            'Data Type': data_type,
+            'Channel': channel,
+            'Charge': value,
+            'Err Abs': value_err
+        })
+
+def remove_bad_peaks(h, x_pos, win):
+    """
+    At each value (`a`) in `x_pos`, fits a gaussian to `h` in the window `(a - win, a + win)`.
+    If the mean of the fitted gaussian is less than or equal to `win * 0.25` away from `a`, then `a` is called a good peak.
+    
+    Returns an array of all the good peaks in `x_pos`.
+    """
+    rm = []
+    for i in range(len(x_pos)):
+        ge = ROOT.TF1("ge", 'gaus', x_pos[i] - win, x_pos[i] + win)
+        ge.SetLineColor(3)
+        if i == 0:
+            r = h.Fit(ge, 'SRQ')
+        else:
+            r = h.Fit(ge, 'SRQ+')
+        print(f'Peak {i} difference: {np.abs(x_pos[i] - ge.GetParameter(1))}')
+        if np.abs(x_pos[i] - ge.GetParameter(1)) > win * 0.25:
+            rm.append(i) 
+    x_pos = np.delete(x_pos, rm)
+    return x_pos
```

Comment:
It looks like this isn't used. Can you remind me what you were using this function for?

Reply:
This was used when we were still using ROOT's peak finding algorithm to fit the SPE histogram. Sometimes it would find a peak that clearly wasn't a peak, so I made this to filter out the "bad peaks". I don't think we'll use the peak finding algorithm again, so I'll delete it.